openquake/calculators/classical.py

```diff
@@ -184,52 +181,37 @@ def gen_args(self, csm, monitor):
-            if oq.optimize_same_id_sources:
-                iterargs = self._args_by_trt(
-                    csm, src_filter, param, num_tiles, maxweight)
+                    csm = self.csm.filter(
+                        SourceFilter(tile, oq.maximum_distance))
             else:
-                iterargs = self._args_by_grp(
-                    csm, src_filter, param, num_tiles, maxweight)
+                csm = self.csm
             num_tasks = 0
             num_sources = 0
+            if monitor.operation == 'pmap_from_grp':
+                iterargs = self._args_by_grp(csm, param, num_tiles, maxweight)
+            else:  # pmap_from_trt
+                iterargs = self._args_by_trt(csm, param, num_tiles, maxweight)
             for args in iterargs:
                 num_tasks += 1
                 num_sources += len(args[0])
                 yield args + (monitor,)
             logging.info('Sent %d sources in %d tasks', num_sources, num_tasks)
         source.split_map.clear()
 
-    def _args_by_grp(self, csm, src_filter, param, num_tiles, maxweight):
-        ngroups = sum(len(sm.src_groups) for sm in csm.source_models)
-        for sm in csm.source_models:
-            for sg in sm.src_groups:
-                gsims = self.csm.info.gsim_lt.get_gsims(sg.trt)
-                if num_tiles <= 1:
-                    logging.info(
-                        'Sending source group #%d of %d (%s, %d sources)',
-                        sg.id + 1, ngroups, sg.trt, len(sg.sources))
-                self.csm.add_infos(sg.sources)
-                if sg.src_interdep == 'mutex':  # do not split the group
-                    sg.samples = sm.samples
-                    yield sg, src_filter, gsims, param
-                else:
-                    for block in self.csm.split_sources(
-                            sg.sources, src_filter, maxweight):
-                        block.samples = sm.samples
-                        yield block, src_filter, gsims, param
-
-    def _args_by_trt(self, csm, src_filter, param, num_tiles, maxweight):
-        for trt, sources in self.dic.items():
-            gsims = self.csm.info.gsim_lt.get_gsims(trt)
-            for block in self.csm.split_sources(
-                    sources, src_filter, maxweight):
-                block.samples = sources[0].samples
-                param['sm_id'] = sources[0].sm_id
-                yield block, src_filter, gsims, param
+    def _args_by_grp(self, csm, param, num_tiles, maxweight):
+        for sg in csm.src_groups:
+            if sg.src_interdep == 'mutex':
+                gsims = csm.info.gsim_lt.get_gsims(sg.trt)
+                self.csm.add_infos(sg.sources)  # update self.csm.infos
+                yield sg, csm.src_filter, gsims, param
+
+    def _args_by_trt(self, csm, param, num_tiles, maxweight):
+        opt = self.oqparam.optimize_same_id_sources
+        for trt, sources in csm.get_sources_by_trt(opt).items():
```

Comment:
Please add a comment about the fact that _args_by_grp considers all sources and it needs to select those that are mutex, whereas _args_by_trt uses get_sources_by_trt that only selects sources that are not mutex